### PR TITLE
Add support for Azure DevOps RunPipelines service endpoint

### LINKED ISSUE #182: [Feature request] Add  support for service endpoint for Azure DevOps
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure DevOps service endpoint is used to start other pipelines.

### New or Affected Resource(s)

* azuredevops_serviceendpoint_runpipeline

### Potential Terraform Configuration

```hcl
resource "azuredevops_serviceendpoint_runpipeline" "serviceendpoint" {
	project_id	      = azuredevops_project.project.id
	service_endpoint_name = "my-pipelines-service-connection"
	organization_name     = "my-organization-name"
	auth_personal {
	# Also can be set with AZDO_PERSONAL_ACCESS_TOKEN environment variable
		personal_access_token = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
	}
	description = "Managed by Terraform"
}
```

### PR DESCRIPTION
**Pull request by zhmurko**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

This pull request introduces a new resource `azuredevops_serviceendpoint_devops` that allows to manage Azure DevOps Service Connection. This connection is used to trigger subsequent builds from Azure Pipelines with `task: RunPipelines@1` yaml definition. 

Fix for Issue Number: #182 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

**Comment by zhmurko** [comment]
build has failed by 60 minutes timeout. need a rebuild or increase azdo job timeout.

**Comment by xuzhang3** [comment]
@zhmurko    Is 'azdoapi' a  new devops extension developed by your team? I cannot find an extension type `azdoapi` . If you means **Azure Repos/Team Foundation Server**, the type should be `externaltfs`
image


Error message:
```log
--- FAIL: TestAccServiceEndpointAzureDevOps_PersonalTokenBasic (24.82s)
    testing.go:673: Step 0 error: errors during apply:
        
        Error: Error creating service endpoint in Azure DevOps: Unable to find service connection type 'azdoapi' using authentication scheme 'Token'.
        
          on /tmp/tf-test450287256/main.tf line 10:
          (source code not available)
```

**Comment by zhmurko** [comment]
hello @xuzhang3 thank you for looking into this issue. 
I was pretty sure that's generally pre-installed type of Service Connection, but it turns out that it's one of Extensions installed from Marketplace manually: [CSE-DevOps.RunPipelines](https://marketplace.visualstudio.com/items?itemName=CSE-DevOps.RunPipelines)
The "azdoapi" is a type of "Azure DevOps Service Connection" as shown on the screenshot below:

The value was picked up from API call to `https://dev.azure.com/my-org-name/_apis/serviceendpoint/endpoints` on save action via UI on Azure DevOps Cloud portal.
We use this service connection to trigger subsequent builds from YAML pipelines and really desire to create such endpoints with Terraform code altogether with the infrastructure.

File: website/docs/r/serviceendpoint_devops.html.markdown [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+## Relevant Links
+
+- [Azure DevOps Service REST API 5.1 - Agent Pools](https://docs.microsoft.com/en-us/rest/api/azure/devops/serviceendpoint/endpoints?view=azure-devops-rest-5.1)
+
+## Import
+
+Azure DevOps Service Endpoint can be imported using the serviceendpoint id (resourceId in your browser's url bar), e.g.
+
+```sh
+ terraform import azuredevops_serviceendpoint_devops.serviceendpoint db0541e6-ae9f-474d-ab83-1f7913839080
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_devops.html.markdown]:
```suggestion
 terraform import azuredevops_serviceendpoint_devops.serviceendpoint projectID/00000000-0000-0000-0000-000000000000`
```

File: website/docs/r/serviceendpoint_devops.html.markdown [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+- `project_id` - The project ID or project name.
+- `service_endpoint_name` - The Service Endpoint name.
+
+## Relevant Links
+
+- [Azure DevOps Service REST API 5.1 - Agent Pools](https://docs.microsoft.com/en-us/rest/api/azure/devops/serviceendpoint/endpoints?view=azure-devops-rest-5.1)
+
+## Import
+
+Azure DevOps Service Endpoint can be imported using the serviceendpoint id (resourceId in your browser's url bar), e.g.
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_devops.html.markdown]:
```suggestion
Azure DevOps Service Endpoint can be imported using the `project id`, `service connection id` , e.g.
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		MaxItems:      1,
+		Elem:          authPersonal,
+		ConflictsWith: []string{},
+	}
+
+	return r
+}
+
+// Convert internal Terraform data structure to an AzDO data structure:
+//
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: Unused code lines can be removed

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+func azdoExpandAuthPersonalSet(d *schema.Set) map[string]string {
+	authPerson := make(map[string]string)
+	val := d.List()[0].(map[string]interface{}) //auth_personal only have one map configure structure
+	authPerson["apitoken"] = val[azdoPersonalAccessToken].(string)
+	return authPerson
+}
+
+// Convert AzDO data structure to internal Terraform data structure
+//
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: Unused code lines can be removed.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+)
+
+const (
+	azdoPersonalAccessToken = "personal_access_token"
+)
+
+// ResourceServiceEndpointAzureDevOps schema and implementation for Azure DevOps service endpoint resource
+func ResourceServiceEndpointAzureDevOps() *schema.Resource {
+	r := genBaseServiceEndpointResource(flattenServiceEndpointAzureDevOps, expandServiceEndpointAzureDevOps)
+	r.Schema["organization"] = &schema.Schema{
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: Can we rename `organization` to `organization_name` or `org_name` ?

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	patHashKey, patHashSchema := tfhelper.GenerateSecreteMemoSchema(azdoPersonalAccessToken)
+	authPersonal.Schema[patHashKey] = patHashSchema
+	r.Schema["auth_personal"] = &schema.Schema{
+		Type:          schema.TypeSet,
+		Optional:      true,
+		MinItems:      1,
+		MaxItems:      1,
+		Elem:          authPersonal,
+		ConflictsWith: []string{},
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: Conflict schema  can be removed.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+				Sensitive:    true,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+			},
+		},
+	}
+	patHashKey, patHashSchema := tfhelper.GenerateSecreteMemoSchema(azdoPersonalAccessToken)
+	authPersonal.Schema[patHashKey] = patHashSchema
+	r.Schema["auth_personal"] = &schema.Schema{
+		Type:          schema.TypeSet,
+		Optional:      true,
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: If I skip the  `auth_personal` configuration, after  execute `terraform apply`, I got error:
```log
Error: Error updating service endpoint in Azure DevOps: Unable to find service connection type 'azdoapi' using authentication scheme 'InstallationToken'.

  on main.tf line 5, in resource "azuredevops_serviceendpoint_devops" "serviceendpoint":
   5: resource "azuredevops_serviceendpoint_devops" "serviceendpoint" {


```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/serviceendpoint"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
+)
+
+const (
+	azdoPersonalAccessToken = "personal_access_token"
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: Duplicate with `personalAccessToken`

File: website/docs/r/serviceendpoint_devops.html.markdown [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+# azuredevops_serviceendpoint_devops
+
+Manages a Azure DevOps Service Connection service endpoint within Azure DevOps. This service connection allows to trigger other subsequent builds with yaml block `task: RunPipelines@1`
+
+## Example Usage
+
+```hcl
+resource "azuredevops_project" "project" {
+  project_name       = "Sample Project"
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_devops.html.markdown]: `project_name` has been renamed to `name` in the latest master

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_devops_test.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+					testutils.CheckServiceEndpointExistsWithName(tfSvcEpNode, serviceEndpointNameSecond),
+				),
+			},
+		},
+	})
+}
+
+// validates that an apply followed by another apply (i.e., resource update) will be reflected in AzDO and the
+// underlying terraform state.
+func TestAccServiceEndpointAzureDevOps_CreateAndUpdate(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_devops_test.go]: This test case is duplicate with `TestAccServiceEndpointAzureDevOps_PersonalTokenUpdate`

**Comment by zhmurko** [comment]
@xuzhang3 thank you for suggestions, I'm going to review and implement them soon

File: website/docs/r/serviceendpoint_devops.html.markdown (hunk 1 above)
Comment (zhmurko) [reply website/docs/r/serviceendpoint_devops.html.markdown]: [x] updated this documentation page

File: website/docs/r/serviceendpoint_devops.html.markdown (hunk 2 above)
Comment (zhmurko) [reply website/docs/r/serviceendpoint_devops.html.markdown]: [x] updated this documentation page

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 3 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: [x] removed the commented examples

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 4 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: [x] removed the commented examples

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 5 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: [x] renamed `organization` to `organization_name` and updated the documentation page

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 6 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: [x] removed `ConflictsWith`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 7 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: added a unit test for `default` settings with empty `auth_personal` personal block

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 8 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: [x] removed this constant value

File: website/docs/r/serviceendpoint_devops.html.markdown (hunk 9 above)
Comment (zhmurko) [reply website/docs/r/serviceendpoint_devops.html.markdown]: [x] renamed `project_name` to `name` in project definition in this example

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_devops_test.go (hunk 10 above)
Comment (zhmurko) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_devops_test.go]: [x] removed this test case and added a test case for default settings

**Comment by zhmurko** [comment]
hi @xuzhang3 
please review the updated pull request, all your suggestions are implemented.

**Comment by zhmurko** [comment]
and one question, does it worth to rename this resource to `azuredevops_serviceendpoint_runpipeline` to avoid confusions?  in the UI this service connection is called just as a "Azure DevOps Service Connection", that may lead to miscommunication.

**Comment by xuzhang3** [comment]
> and one question, does it worth to rename this resource to `azuredevops_serviceendpoint_runpipelines` to avoid confusions? in the UI this service connection is called just as a "Azure DevOps Service Connection", that may lead to miscommunication.

I would suggest rename with `azuredevops_serviceendpoint_pipelinerunner` or `azuredevops_serviceendpoint_pipelinedelegator`,  this extension is kind like the admin of another pipeline.

**Comment by zhmurko** [comment]
so I will refactor the code and docs with `azuredevops_serviceendpoint_pipelinerunner`  resource name.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	authPerson := make(map[string]string)
+	if len(d.List()) == 1 {
+		val := d.List()[0].(map[string]interface{}) //auth_personal block may have only one element inside
+		authPerson["apitoken"] = val["personal_access_token"].(string)
+	}
+	return authPerson
+}
+
+func azdoPersonalAccessTokenField() *schema.Resource {
+	field_name := "personal_access_token"
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: variable name style should follow camel style

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+		MinItems: 1,
+		MaxItems: 1,
+		Elem:     azdoPersonalAccessTokenField(),
+	}
+
+	return r
+}
+
+// Convert internal Terraform data structure to an AzDO data structure:
+func expandServiceEndpointAzureDevOps(d *schema.ResourceData) (*serviceendpoint.ServiceEndpoint, *string, error) {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: function `expandServiceEndpointAzureDevOps` never throw an error, I would suggest remove the `error ` from the return parameter.

**Review by xuzhang3 — COMMENTED** [review]
Few changes requested but otherwise looks good.  Can we rename function name prefix `azdo`  to another name like `pipelineRunner` or something else, it's a bit odd with prefix `azdo`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go (hunk 12 above)
Comment (zhmurko) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_devops.go]: done

**Comment by zhmurko** [comment]
hi @xuzhang3 
the refactoring has been completed. The build now fails again on `Unable to find service connection type 'azdoapi'` message, and I believe it's related to disabled/removed plugin `RunPipelines` from AzureDevOps. Could you please enable it so acceptance test can pass.

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**